virttest/qemu_vm.py

```diff
@@ -1330,9 +1330,20 @@ def sort_key(dev):
             for pcic in params.objects("pci_controllers"):
                 dev = devices.pcic_by_params(pcic, params.object_params(pcic))
                 pcics.append(dev)
+            pcie_extra_root_port = params.get('pcie_extra_root_port', 0)
+            for num in range(int(pcie_extra_root_port)):
+                pcics.append(devices.pcic_by_params("pcie_root_port_%s" % num,
+                                                    {"type": "pcie-root-port"}))
             if params.get("pci_controllers_autosort", "yes") == "yes":
                 pcics.sort(key=sort_key, reverse=False)
-            map(devices.insert, pcics)
+            from virttest.qemu_devices.utils import DeviceError
+            for pcic in pcics:
+                try:
+                    devices.insert(pcic)
+                except DeviceError as err:
+                    if pcic.get_param("driver") == "pcie-root-port" and str(err).find(
+                                    "Failed to insert device:\n%s" % pcic.str_long()):
```

Comment:
Oh the message was suppose to be here. The `str(err).find(...)` is always True as on no match it returns `-1`. Please use `"Failed ..." % pcic.str_long() in str(err)`.

Reply:
Actually I looked at this closely and it can't ever work properly the way I meant it to, so you can just remove it. The problem is that controllers are added soon in the code, so even if we wanted to add the devices later it'd be impossible.

I'm sorry for realizing this only now...

Reply:
(I mean if the test wants to occupy all the PCI slots, it has to manually set `pci_extra_root_port=0`, otherwise those extra controllers will occupy the slots that are meant for the test device and skipping inserts won't help because the extra slots are added early in the code. It'd work if we moved the extra controllers at the end of `make_create_command` function, but I don't think it's worth the effort. So please just use the `devices.insert` and eventually we'll find tests that use too much PCI slots and we'll have to manually treat their configuration...

Reply:
... alternatively we could define the `pcie_extra_root_port = 3` only for hotplug cases. Maybe we should consider that, what do you think? (instead of breaking other test simply start adding this to tests that require the extra root ports. I think that'd be even better)

Reply:
@ldoktor Yes that's actually not very useful since other device will be inserted later that extra root port. As far as I know, we should have more hotplug test cases than `pci_extra_root_port=0` cases, so I still prefer set it to 3 by default, and also because up layer tools will boot up q35 machine with several empty root port, I think it would be better to align to them.

Reply:
I also discussed with q35 feature owner @jingzhao84 , we think minor cases will be affected by these extra root port. So I will only remove the exception part if you can accept keeping `pcie_extra_root_port = 3` as default configuration. And we can set `pci_extra_root_port=0` for possible incompatible cases later if we hit any. What do you think?

Reply:
Sure, there are no questions about removing the `exception`. As for the default, yes. I'd prefer 0, but I have nothing against 3, especially when the upper layers follow the same rule.